### DISCUSSION TITLE
PlgEditors-xtd - class name problem

### DISCUSSION THREAD

**Discussion by linpi1** [joomengine/Joomla-Component-Builder#809] [Extention Issues]
I want to create an button as an editor extension.  In the php file, the default class name of the plugin starts with PlgEditors-xtd + the plugin name, example class PlgEditors-xtdBookshopButton extends CMSPlugin. This dash in the class name makes an error in PHPStorm and also when I open an article: `syntax error, unexpected '-', expecting '{' `
Therefore I changed the class name to  PlgEditors_xtdBookshopButton. No error anymore.
But I cant not just change it with `/***[REPLACE<>$$$$]***/` this make a compilation warning:
`Custom Code Warning
File plugin_editors-xtd_bookshopbutton/bookshopbutton.php could not be found, so the custom code for this file could not be addded.`
So what to do know?

Another thing, the expected button doesn't appear in the editor. I compared the code (function onDisplay($name)) with editor-xtd/article.php and it is almost the same, but still no button. Is there something missing to add ?

**Comment by linpi1** [comment]
ok it start to work.
manually I replaced the class name to PlgButton + plugin name. 
in function onDisplay I replaced Session::getFormToken() with JSession::getFormToken() 
and $button = new CMSObject; with $button = new JObject;  (here I have to see why, since JObject is deprecated)
and $button->text = Text::_('PLG_EDITOR... with $button->text = JText::_('PLG_EDITOR... (just a typo)
know the button is shown.

**Comment by linpi1** [comment]
of course `Session::getFormToken()` & `$button = new CMSObject;` were  perfectly fine,
I had just to add some Class Header:
`use Joomla\CMS\Object\CMSObject;
use Joomla\CMS\Session\Session;`

**Reply by Llewellynvdm** [reply]
There are plugins in JCB that allow you to [add class headers](https://www.youtube.com/watch?v=TjoH5rj4UgM&list=PLQRGFI8XZ_wsV6PDFXd21B7O0X70JlNd3&index=7) to parts of the component via the GUI.

**Comment by linpi1** [comment]
yes, I already did it on this way. 
the only problem left is the wrong class name `PlgEditors-xtdBookshopButton` instead of `PlgButtonBookshopButton` which I  correct manually, since` /***[REPLACE<>$$$$]***/` doesn't work, respectively make a compilation warning. I think it has with the file path to do, since the path of plugin is outside of the path of the component. just a guess. 
I can try to find where to change in the JCB compiler.

**Reply by Llewellynvdm** [reply]
I am wondering if this is not a possible bug, sorry I did not ready your first post carefully the first time. This `PlgEditors-xtd` name comes from the [boilerplate](https://github.com/vdm-io/boilerplate/tree/jcb/plugins) and is added to this table `/administrator/index.php?option=com_componentbuilder&view=Joomla_plugin_groups` so you can edit the name and see if it fixes the issue.

**Reply by Llewellynvdm** [reply]
We may need to add an extra layer of naming. Since the folder as a **group** has this `-` but not the [class](https://github.com/vdm-io/boilerplate/blob/jcb/plugins/editors-xtd/foo.php#L25).

**Reply by Llewellynvdm** [reply]
Yoiu will see the [other classes](https://github.com/vdm-io/boilerplate/blob/jcb/plugins/finder/foo.php#L23) use the group name in the class name, except this one. So I think we have a bug.... and we need to fix this.

**Reply by linpi1** [reply]
when I in `/administrator/index.php?option=com_componentbuilder&view=Joomla_plugin_groups` and editors-xtd change the name to `Button` the class name PlgButton+pluginname is is ok, but in the xml file group becomes `<extension type="plugin" version="3.8" group="button" method="upgrade">` instead of `<extension type="plugin" version="3.8" group="editors-xtd" method="upgrade">` .
it is for  editors-xtd definitely a special case for the class name.
